Custom Wayland compositor

ABeltramo opened the pull request:
# TODO

- [x] run wayland compositor outside gstreamer in order to support pause/resume
- [x] avoid flooding of IDR requests
- [x] remove thread pool
- [x] add proper startup script to the Dockerfile
  - [x] Download additional Nvidia drivers if required
- [x] Allow changing the docker-socket path via ENV 
- [x] Change `config.toml`
  - [x] Add all current GOW apps as default apps
  - [x] Change the gstreamer pipelines to be a hierarchy of encoders (ex: `nvenc -> vaapi -> software`) so that we can automatically plug the right one at runtime
  - [x] Add ability to change the render node per app

ABeltramo committed:
feat: basic wayland docs, updated Sunrise

Drakulix committed:
feat: Add waylanddisplaysrc plugin

Drakulix committed:
docker: Add wayland plugin to Dockerfile

Drakulix committed:
feat: Add wayland plugin to default config

Drakulix committed:
docker: Add nvidia build arg

Drakulix committed:
docs: Clarify zero-copy situation

Drakulix committed:
feat: wayland: Use correct age

Drakulix committed:
feat: wayland: Implement frame pacing

Drakulix committed:
feat: wayland: Add support for absolute mouse movement

Drakulix committed:
feat: wayland: properly handle client cursors

ABeltramo committed:
fix: Dockerfile nvidia URL

ABeltramo committed:
feat: plugged run command with wayland and xorg sockets

ABeltramo committed:
feat: sending created devices paths back to the plugin

Drakulix committed:
feat: wayland: handle input-device events

ABeltramo committed:
fix: wayland scroll support, slow down RTSP in order to warm up the gstreamer pipeline

ABeltramo committed:
fix: creating virtual inputs even when there's no cmd

ABeltramo committed:
fix: immutable app state, code cleanup

Drakulix committed:
feat: wayland: Hide cursor after 5 sec

ABeltramo committed:
feat: multi-user over the same ports

Added support for pause, resume and stop.
Major code refactor, simplified, removed unnecessary events.

fixes #22

ABeltramo committed:
fix: tests

Drakulix committed:
feat: wayland: Add shortcuts

Drakulix committed:
uinput: Add udev rules

ABeltramo committed:
feat: pulseaudio automate virtual sink creation

ABeltramo committed:
fix: missing libpulse dependency

ABeltramo committed:
feat: implemented Docker API client

ABeltramo committed:
feat: automated PulseAudio start/stop via Docker

ABeltramo committed:
refactor: better decoupled run process with video/audio streaming

ABeltramo committed:
refactor: abstract Runner class

ABeltramo committed:
feat: added running Docker apps via config.toml

ABeltramo committed:
fix: github actions missing libcurl

ABeltramo committed:
fix: docker build

Drakulix committed:
wayland: Catch panics

Drakulix committed:
wayland: Cache EGLDisplays (and allow software rendering)

Drakulix committed:
wayland: Manually implement wl_drm

Drakulix committed:
wayland: Giant rewrite

ABeltramo committed:
feat: added uinput in order to enumerate children of devices

Drakulix committed:
wayland: Move src

ABeltramo committed:
feat: Docker basic retro user with 1000:1000

ABeltramo committed:
feat: docker add support to override the default `create` JSON post request

Drakulix committed:
wayland: Don't mangle c-api

Drakulix committed:
wayland: Don't throttle frame updates

Drakulix committed:
wayland: properly null-terminate c-api strings

Drakulix committed:
wayland: Forget about buffers

ABeltramo committed:
feat: pause/resume support, driving Wayland compositor outside Gstreamer

ABeltramo committed:
fix: GA tests

ABeltramo committed:
feat: added ENV variables

ABeltramo committed:
fix: better use of smart pointers to avoid exception on resume

ABeltramo committed:
refactor: boost ASIO and docker run

Drakulix committed:
docker: Fixup podman build

ABeltramo committed:
feat: basic IDR backoff strategy

Drakulix committed:
streaming/video: Set frame-type

ABeltramo committed:
fix: tests

ABeltramo committed:
fix: properly sync packets with Moonlight

ABeltramo committed:
fix: tests

ABeltramo committed:
feat: config.toml V2

 - Added automatic encoder selection
 - Added `run_gid` and `run_uid`
 - Added boolean start_virtual_compositor

Drakulix committed:
wayland: Drop Xwayland

Drakulix committed:
wayland: Resize windows on resolution change

ABeltramo committed:
fix: tests, removed hack for XWayland

Drakulix committed:
wayland: Only send frame events after rendering

ABeltramo committed:
fix: tests

ABeltramo committed:
fix: catching JSON parse exception, avoids unnecessary double g_source_remove

ABeltramo committed:
feat: better encoder auto-selector based on Gstreamer available elements

ABeltramo committed:
fix: passing caps to appsrc, killing container when can't stop nicely

ABeltramo committed:
fix: appsrc live support, added PTS, DTS and Duration

ABeltramo committed:
fix: allow compilation on non Linux platforms

ABeltramo committed:
feat: implemented appsink with boost::asio instead of using custom gstreamer plugin

Drakulix committed:
wayland: Add support for wp_presentation

Drakulix committed:
uinput: Don't advertise force-feedback (yet)

Drakulix committed:
uinput: Allocate four controllers in advance

Drakulix committed:
wolf: Expose display_mode as env vars for gamescope

Drakulix committed:
uinput: Store previous controller state per controller id

ABeltramo committed:
fix: better stacktrace on exception

ABeltramo committed:
fix: tests

ABeltramo committed:
feat: better handling of XDG_RUNTIME_DIR updated config file

ABeltramo committed:
docs: basic instructions

ABeltramo committed:
fix: VAAPI default pipeline

ABeltramo committed:
CI: Added automated linter

ABeltramo committed:
Committing clang-format changes

ABeltramo committed:
docs: updated configuration

ABeltramo committed:
feat: added support for aarch target

ABeltramo committed:
fix: removed boost thread pool

ABeltramo committed:
revert: back to our custom plugin instead of appsink

ABeltramo committed:
feat: adding amfcodec to gstreamer

ABeltramo committed:
Committing clang-format changes

ABeltramo committed:
fix: cargo target build when not defined

Drakulix committed:
wayland: support relative pointer events

Drakulix committed:
wayland: Implement pointer constraints

ABeltramo committed:
feat: docker API added option to change socket path via ENV and via Config file

ABeltramo committed:
fix: smart pointers and overlapping sessions

Using unique_ptr for our custom wayland display to make sure that's correctly destroyed when the session is over

Refactored RTP Ping and Gstreamer pipelines start to avoid double calling and reference holding;
Moonlight keeps pinging during the whole duration of the streaming, we have to make sure that we only get one ping to start the pipelines; this should be the case now.

ABeltramo committed:
fix: uint32 as the audio sequence number

ABeltramo committed:
Committing clang-format changes

Drakulix committed:
Don't manage pulse for provided servers

Drakulix committed:
Revert "wayland: Implement pointer constraints"

This reverts commit bb7feb4efe0425aca4d35bcb148e5fb000d95835.

ABeltramo committed:
docs: docker compose, ports, udev rules explanation

Drakulix committed:
Revert "fix: uint32 as the audio sequence number"

This reverts commit f3f9cc000f98f79fbf363ba3066f4f8997f9609a.

ABeltramo committed:
feat: QuickSynk support, manual driver build

ABeltramo committed:
feat: Intel QuickSync support, new base image, fix gstreamer dependencies and image cleanup

ABeltramo committed:
feat: moved docker files, better use of gow/base-app

ABeltramo committed:
docs: better Nvidia support and instructions

ABeltramo committed:
docs: added a few more critical bits

ABeltramo committed:
docs: fix docker-compose for Nvidia

ABeltramo committed:
feat: deal with Nvidia extra paths in Wolf

ABeltramo committed:
feat: Docker session based folder separation

fixes #25

ABeltramo committed:
feat: added Docker logs on exit

ABeltramo committed:
Committing clang-format changes

Drakulix commented on src/wolf/wolf.cpp:
```diff
@@ -246,6 +246,19 @@ auto setup_sessions_handlers(const immer::box<state::AppState> &app_state,
+            full_devices.push_back("/dev/nvidia-uvm");
+            full_devices.push_back("/dev/nvidia-uvm-tools");
+            full_devices.push_back("/dev/nvidiactl");
+          }
```
Just because you are encoding with an nvidia devices, you don't have to render on an nvidia device. Also you might have multiple nvidia devices (and need to map the corresponding `/dev/nvidia*` device).

Can we please not force this but make it optional? Or infer this information from the render-node for the wayland compositor instead?

ABeltramo replied:
Both very valid points, I knew I had to move it behind a flag or something.  
The idea is to try to avoid having users to mess with the config file and provide some sensible default, in this case, since we know that nvidia is working we can safely add those.

I agree with your point, though. I'll think of a better way to (de)activate this mode!

Drakulix replied:
I guess this plays into being able to set a custom render-node per app? So we could use the encoding-device as a default and overriding the render-node would disable automatic mapping or invoke different logic to determine the vendor of the node.

ABeltramo committed:
fix: nvidia with multi GPU support

Added libpci and libdrm in order to properly map the render-node with the corresponding primary node and additional nvidia specific devices.

ABeltramo committed:
feat: added support for custom render-node per app

QuaiGoner reviewed (commented):
Shouldnt volume mapping for persistent user storage be:
```
~wolf/storage:/etc/wolf
~wolf/cfg/:/wolf/cfg
```
and not
`/etc/wolf/wolf:/wolf/cfg`?

ABeltramo committed:
fix: compile on non-linux system

Moved default images from `wolf-changes` to `edge`

ABeltramo committed:
docs: easier quickstart, fixed links after merging, added pages

ABeltramo committed:
docs: multi GPU, base troubleshooting
